run functorch decomps after functionalization when enabled

bdhirsh opened the pull request:
This is a short-to-midterm fix for https://github.com/pytorch/pytorch/issues/83923. By running functionalization before decomps, we guarantee that functionalization won't have to see any primtorch view/inplace ops like `broadcast_in_dim`.

This will only really be a problem if there's a function in the decomposition table that decomposes a functional op into mutations. If that comes up later, we'll need to revisit https://github.com/pytorch/pytorch/issues/83923.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83993
* __->__ #83992

bdhirsh committed:
get prims::broadcast_in_dim to work in functionalization

[ghstack-poisoned]

bdhirsh committed:
Update on "get prims::broadcast_in_dim to work in functionalization"

Fixes https://github.com/pytorch/pytorch/issues/83923

I only bothered with `broadcast_in_dim` for now, but we might need to add other view ops later





[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -2028,15 +2028,22 @@ TEST(NewOperatorRegistrationTest, testImplNoDefGetsCaught) {
       error_str += "\n";
   }
-  ASSERT_EQ(danglingImpls.size(), 0) << error_str;
+  ASSERT_EQ(danglingImpls.size(), dangling_prim_ops) << error_str;
```
sigh

ezyang reviewed (approved):
If we need to do more of these, let's expose enough API that these can be written in Python

Chillee commented:
I'm kind of against doing this, tbh. I feel like this can be resolved by just doing decompositions after functionalization.

bdhirsh committed:
Update on "get prims::broadcast_in_dim to work in functionalization"

Fixes https://github.com/pytorch/pytorch/issues/83923

I only bothered with `broadcast_in_dim` for now, but we might need to add other view ops later





[ghstack-poisoned]

bdhirsh commented:
Updated based on suggestions in that linked issue - now we'll just run decomps after functionalization.

I confirmed that with the repro command the errors also goes away.
